tycho-extras/tycho-pomless/src/main/java/org/eclipse/tycho/pomless/PomXMLMapping.java

```diff
@@ -0,0 +1,37 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2022 Hannes Wellmann and others.
+ * 
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ * 
+ * Contributors:
+ * Hannes Wellmann - initial API and implementation based on org.sonatype.maven.polyglot.mapping.XmlMapping
+ *******************************************************************************/
+package org.eclipse.tycho.pomless;
+
+import java.io.File;
+import java.util.Map;
+
+import org.codehaus.plexus.component.annotations.Component;
+import org.sonatype.maven.polyglot.mapping.Mapping;
+import org.sonatype.maven.polyglot.mapping.MappingSupport;
+
+@Component(role = Mapping.class, hint = "pom.xml") //TODO: adjust hint?
```

Comment:
Why do we need this? With the new priority should not TychoMappings be asked anyways before? The check for pom.xml must be done in the tycho mappings anyways...

Reply:
The current suggestion is to have to following mapping priority:

1. pom.xml
2. bundle
3. feature
4. repository
5. target
6. aggregator
7. XMLMapping

If we don't have a dedicated PomXMLMapping then the mapping priority would have to be
1. bundle
2. feature
3. repository
4. target
5. XMLMapping
6. aggregator

But we have to keep the check in `AbstractTychoMapping.locatePom()` if a `pom.xml` is already present in the project's root.
So while the `XMLMapping` has a low priority (in order to handle `feature.xml` files), the pom.xml files it is intended to handle have actually maximum priority. I found that odd and therefore added the PomXMLMapping so that the priority of the mappings reflects the priority of the files they handle, in order to keep that model (IMO) simpler to understand.

But of course it would also work the other way. It is more a question what we prefer.

> The check for pom.xml must be done in the tycho mappings anyways...

Why must the check be done anyways if we have a dedicated PomXMLMapping?
At least there is not test that fails with the current suggestion. But if you can think of a scenario were it is required some tests should be added.

Reply:
> At least there is not test that fails with the current suggestion.

I don't think we have much test-coverage for the pomless stuff :-\

> Why must the check be done anyways if we have a dedicated PomXMLMapping?

Because `TychoMappings` should be ignored if there is a `pom.xml`, they should not be ignored if there is any other mapping with higher priority (what is what you have implemented here), even though the visible effect is the same in most scenarios this is a different semantics.

> I found that odd and therefore added the PomXMLMapping so that the priority of the mappings reflects the priority of the files they handle,

We don't really want or should handle `pom.xml` files so adding a mapping in in Tycho is simply wrong (and maybe even more odd), the choice that `pom.xml` always takes precedence is a Tycho choice but not normative at all when it comes to polyglot in general.

Reply:
> > At least there is not test that fails with the current suggestion.
> 
> I don't think we have much test-coverage for the pomless stuff :-\

At least with the pomless-model tests we have some more but of course there could be more.

> > I found that odd and therefore added the PomXMLMapping so that the priority of the mappings reflects the priority of the files they handle,
> 
> We don't really want or should handle `pom.xml` files so adding a mapping for it in Tycho is simply wrong (and maybe even more odd), the choice that `pom.xml` always takes precedence is a Tycho choice but not normative at all when it comes to polyglot in general.
> 
> Just assume I also want to use another mapping that do not has this rule, then Tycho can't be used without giving a higher priority than our "pom xml mapping" just because we don't handle this rule internally. You should always be aware that marking anything `@Component` it makes it visible to the whole maven world.

Yes, that's indeed a reason against the first suggestion. Although I think it wouldn't be a bad choice for other polyglot-languages as well we must not impose this rule from Tycho. I removed the PomXMLMapping again and changed the priorities to:

1. bundle
2. feature
3. repository
4. target
5. XMLMapping
6. aggregator